### ISSUE TITLE
Add splines for covariant B components

### ISSUE THREAD

**Issue by krystophny** [itpplasma/libneo#34]

**Comment by krystophny** [comment]
*Automated triage pass (Claude), grounded in the repository at `fe3f76e`. Verify before acting.*

The body is empty, so the scope below is reconstructed from the code and from what was being worked on when the issue was filed (2024-01-31, immediately after `8b7e7de` "32 add boozer fourier transformations (#50)", which added the `efit_to_boozer` Python/Fourier layer). Read as: *the symmetry-flux/Boozer magdata layer splines geometry but not the covariant components of **B**, so every consumer re-derives them by hand.* That gap is still open in the EQDSK path. It was closed independently in the VMEC path in June 2026, which gives us a reference implementation to copy rather than a design to invent.

## Purpose

Guiding-centre drift kinetics in Boozer coordinates needs the covariant components `B_s`, `B_theta`, `B_phi`, not just `|B|` and the metric: the magnetic drift and the parallel dynamics enter through `B_theta = I(s)`, `B_phi = G(s)` and the Jacobian factor `(G + iota*I)`, and `B_s` (nonzero for a finite-beta or non-circular equilibrium) enters `curl h` and hence `nabla s` and the radial drift. Today the EQDSK→Boozer path in libneo supplies none of them from splines; consumers either recompute them from `field_eq` plus a hand-written metric, or approximate them in Python. Two independent reconstructions of the same quantity already disagree in-tree (see below), which is exactly the failure mode a single splined source of truth removes. libneo#305 and its consumers (itpplasma/rabe#98, itpplasma/NEO-RT#44) all want covariant components out of a libneo Boozer evaluator.

## Current state

**Symmetry-flux magdata splines only geometry.** `spline_magdata_in_symfluxcoord` (`src/efit_to_boozer/spline_and_interpolate_magdata.f90:3-69`) periodic-splines exactly five quantities over `theta`:

```fortran
    splcoe(0,1:ntheta)=R_spl(0,1:ntheta,i)      ! ... then Z_spl, bmod_spl,
    splcoe(0,0)=splcoe(0,ntheta)                !     sqgnorm_spl, Gfunc_spl
    call spl_per(nspl,nthetap1,h_theta,splcoe)
```

The backing arrays are declared in `src/efit_to_boozer/efit_to_boozer_mod.f90:13`:

```fortran
    double precision, dimension(:,:,:), allocatable :: R_spl,Z_spl,bmod_spl,sqgnorm_spl,Gfunc_spl
```

`magdata_in_symfluxcoord_ext` (`src/efit_to_boozer/spline_and_interpolate_magdata.f90:73-206`) therefore returns `q, dq_ds, C_norm, dC_norm_ds, sqrtg, bmod, dbmod_dtheta, R, dR_ds, dR_dtheta, Z, dZ_ds, dZ_dtheta, G, dG_ds, dG_dtheta` — no `B_s`, `B_theta`, `B_phi`. The f2py wrapper `magdata` (`src/efit_to_boozer/f2py_interfaces/f2py_efit_to_boozer.f90:106-117`) mirrors that list.

**The consumers each improvise.** `app/efit_to_boozer.f90:145-149` reconstructs the flux functions from a single `field_eq` sample at the last `theta` of the loop:

```fortran
! iota, poloidal and toroidal covariant components:
    aiota=1.d0/q
    B_phi=Bp*R
    B_theta=q*(C_norm-B_phi)
    sqrtg00=(B_phi+B_theta/q)*real(sum(calEm_times_Jb/B_oft**2))*oneovernt
```

i.e. it defines `C_norm = B_phi + iota*B_theta = G + iota*I`. The Python chartmap builder disagrees: `python/libneo/eqdsk_to_boozer_chartmap.py:242` sets `B_phi_cgs = C_norm_arr` under the comment `B_phi = G(s) = R*B_tor = C_norm from magdata`, which drops the `iota*I` term (a systematic `O(iota*I/G)` error in `G`, ~1% for a tokamak), and then rebuilds `B_theta` from scratch by an Ampère loop integral over a `RectBivariateSpline` of the g-file `psi` map (`python/libneo/eqdsk_to_boozer_chartmap.py:245-273`). Both routes exist only because the Fortran side does not spline and expose the covariant components.

**The VMEC path already does this.** `src/boozer_converter.F90` splines `B_theta`, `B_phi` radially (`bcovar_tp_batch_spline`, line 57) and `B_r` on the full 3D Boozer grid behind the `use_B_r` switch (line 38, `br_grid`, `field3d_batch_spline`). The symmetry-flux → Boozer radial transform is `compute_br_from_symflux` (`src/boozer_converter.F90:1034-1076`):

```fortran
                br_grid(i_rho, :, i_phi) = &
                    2.0_dp*rho_tor(i_rho)*Bcovar_symfl(1, i_rho, :, i_phi) &
                    - matmul(coef(1, :)*aiota_arr(ibeg:iend), Gfunc(ibeg:iend, :, i_phi)) &
                    *Bcovar_symfl(2, i_rho, :, i_phi) &
                    - matmul(coef(1, :), Gfunc(ibeg:iend, :, i_phi)) &
                    *Bcovar_symfl(3, i_rho, :, i_phi)
```

This is eq. `(B_r_expl)` of `doc/efit_to_boozer/EFIT_to_Boozer_and_Hamada_converter.tex:241-263` — a derivation written for the EFIT converter but implemented only for VMEC. `test/source/test_boozer_converter_vs_simple.f90:120-146` covers it (it builds `hcovar` and `hcurl` from `B_r`).

**The chartmap interchange format carries only the 1D pieces.** `src/boozer_chartmap.f90:267-270` defines `B_theta(rho)` and `B_phi(rho)`; `src/field/boozer_chartmap_io.f90:92-100` and `src/boozer_chartmap_types.f90:29-30` read them back. There is no `B_r` variable, so a chartmap consumer cannot get the radial covariant component at all without re-running the converter with `use_B_r = .true.`.

## Reproduction / evidence

Build and run the existing EQDSK→Boozer chartmap test, which is the shortest path that exercises all of the above:

```bash
cd /path/to/libneo
make                       # cmake --preset default && cmake --build --preset default
pip install -e ".[dev]"
python -m pytest python/tests/test_eqdsk_to_boozer_chartmap.py -vv
```

The fixture chain is real and committed: `python/tests/efit_to_boozer.inp` (nstep=3600, nlabel=500, ntheta=500, nsurf=2000, mpol=12), `python/tests/field_divB0.inp` pointing at `test/resources/input_efit_file.dat`.

Observed: `test_covariant_components_tokamak_ordering` (`python/tests/test_eqdsk_to_boozer_chartmap.py:76-93`) and `test_B_theta_matches_plasma_current` (`:96-110`) pass, but only because `B_theta` was computed in Python from the g-file, not from the Boozer transform. Expected after this issue: both values come from `magdata`, and a new consistency check `C_norm == B_phi + B_theta/q` holds to `1e-10` relative — today that identity is definitionally true in `app/efit_to_boozer.f90` and *violated* by `eqdsk_to_boozer_chartmap.py:242`, which is the observable defect.

Minimal Python probe of the missing API:

```python
from _efit_to_boozer import efit_to_boozer   # run in a dir with the two .inp files
efit_to_boozer.init()
out = efit_to_boozer.magdata(1, 0.5, 0.0, 0.0)
print(len(out))   # 16 today: no B_s, B_theta, B_phi anywhere in the tuple
```

## Behaviour (BDD)

```gherkin
Scenario: magdata returns covariant components in symmetry flux coordinates
  Given test/resources/input_efit_file.dat via python/tests/field_divB0.inp
    And python/tests/efit_to_boozer.inp with nlabel=500, ntheta=500
  When magdata_in_symfluxcoord_ext is called with inp_label=1, s=0.5, theta=1.0
    and the optional outputs Bcov_s, Bcov_theta, Bcov_phi are present
  Then Bcov_phi agrees with R*Bp from a direct field_eq(R,0,Z) call to 1e-10 relative
    And Bcov_theta agrees with Br*dR_dtheta + Bz*dZ_dtheta to 1e-8 relative
    And Bcov_s agrees with Br*dR_ds + Bz*dZ_ds to 1e-8 relative

Scenario: Boozer-frame covariant components satisfy the Jacobian identity
  Given the same equilibrium and s in {0.2, 0.5, 0.8}
  When bcovar_boozer_from_symflux returns B_theta_B(s), B_phi_B(s)
  Then abs(C_norm - (B_phi_B + B_theta_B/q)) < 1e-10 * abs(C_norm)

Scenario: B_theta from splines matches the enclosed plasma current
  Given test/resources/input_efit_file.dat with Ip read by libneo.eqdsk_base.read_eqdsk
  When B_theta is taken from magdata at the outermost stored surface
  Then abs(B_theta) is within [0.7, 1.05] * 0.2 * abs(Ip) in G*cm
    And python/tests/test_eqdsk_to_boozer_chartmap.py::test_B_theta_matches_plasma_current still passes

Scenario: the chartmap writer stops fabricating B_phi
  Given libneo.eqdsk_to_boozer_chartmap.write_chartmap on the same g-file
  When the produced .chartmap.nc is read back
  Then B_phi(rho) equals the splined Boozer B_phi, not C_norm
    And max|B_phi_new/B_phi_old - 1| is of order iota*I/G, i.e. between 1e-3 and 5e-2
```

## Implementation plan

1. `src/efit_to_boozer/efit_to_boozer_mod.f90` — add to line 13 the arrays `Bcov_s_spl, Bcov_theta_spl` with the same `(0:nspl,0:ntheta,nlabel)` shape, and to line 10 the Lagrange work arrays `Bcov_s_lag, Bcov_theta_lag(nplag)`. Add `Bcov_phi(:)` (allocatable, size `nlabel`) — the toroidal covariant component is a flux function in axisymmetry.
2. `src/efit_to_boozer/spline_and_interpolate_magdata.f90` — add `subroutine compute_bcovar_symfluxcoord` running **after** the existing geometry spline pass. For each `(i=1..nlabel, it=1..ntheta)` call `magdata_in_symfluxcoord_ext` for `R, Z, dR_ds, dR_dtheta, dZ_ds, dZ_dtheta`, call `field_eq(R, 0.d0, Z, Br, Bp, Bz, ...)` (`src/magfie/field_divB0.f90`), and set `Bcov_s = Br*dR_ds + Bz*dZ_ds`, `Bcov_theta = Br*dR_dtheta + Bz*dZ_dtheta`, `Bcov_phi(i) = R*Bp`. Then `spl_per` the two theta-dependent grids exactly as at lines 22-55. Call it from the end of `spline_magdata_in_symfluxcoord` (line 68).
3. Same file — extend `magdata_in_symfluxcoord_ext` (line 73) with `optional, intent(out) :: Bcov_s, Bcov_theta, Bcov_phi, dBcov_s_dtheta, dBcov_theta_dtheta`, evaluated by the identical Horner + `plag_coeff` pattern used for `bmod_lag` at lines 165-190. Optional arguments keep `app/efit_to_boozer.f90`, `app/hamada_to_boozer.f90` and the f2py wrapper compiling unchanged, which matters for the downstream gate (`ci/downstreams`). MEPHIT and GORILLA carry private forks of this routine (`MEPHIT/src/magdata_in_symfluxcoord.f90`, `GORILLA/SRC/magdata_in_symfluxcoord.f90`) and are unaffected.
4. Same file — add `subroutine bcovar_boozer_from_symflux(s, theta, B_s_B, B_theta_B, B_phi_B)` implementing eq. `(B_r_expl)` / `(B_theta)` of `doc/efit_to_boozer/EFIT_to_Boozer_and_Hamada_converter.tex:241-290`. Port the algebra verbatim from `src/boozer_converter.F90:1034-1076` (`compute_br_from_symflux`), substituting the symmetry-flux `G` from `Gfunc_spl` and `dG_ds` from `magdata_in_symfluxcoord_ext` for the VMEC `Gfunc` array and its `plag_coeff` radial derivative. Do **not** duplicate the formula: if the two paths converge, lift it into a shared helper later under libneo#305.
5. `src/efit_to_boozer/f2py_interfaces/f2py_efit_to_boozer.f90` — add the new module arrays to the `use efit_to_boozer_mod, only:` list (lines 3-5), allocate them in `init` next to line 38, deallocate in `deinit`, and extend the `magdata` wrapper (lines 106-117) with the three new outputs. Add a second wrapper `magdata_boozer` for step 4. No CMake change is needed: `src/efit_to_boozer/CMakeLists.txt:65-70` already lists `spline_and_interpolate_magdata.f90` in the `efit_to_boozer` target, and the f2py custom command at the bottom of that file already passes it to `numpy.f2py`.
6. `python/libneo/eqdsk_to_boozer_chartmap.py:232-273` — delete the `RectBivariateSpline`/Ampère block and `B_phi_cgs = C_norm_arr`; take `B_theta_cgs[k]` and `B_phi_cgs[k]` from the extended `magdata` call already made in the loop at lines 225-232. Keep the units comment (CGS `G*cm`).
7. New test `test/source/test_bcovar_symflux.f90`, registered in `test/CMakeLists.txt` with the `add_executable(... ) / target_link_libraries(... efit_to_boozer ${COMMON_LIBS}) / add_test(...)` idiom used at lines 123-136, run in a working directory holding copies of `python/tests/efit_to_boozer.inp` and a `field_divB0.inp` pointing at `test/resources/input_efit_file.dat`. Assert scenarios 1 and 2 above with `error stop`.
8. `python/tests/test_eqdsk_to_boozer_chartmap.py` — add `test_C_norm_equals_G_plus_iota_I` and tighten `test_covariant_components_tokamak_ordering` to compare against the splined values.
9. **Follow-up, separate PR (split this off as its own issue):** add a 3D `B_r(rho,theta,zeta)` variable to the chartmap format — `src/boozer_chartmap.f90:267-303` (`def_var`/`put_var`), `src/boozer_chartmap_types.f90:29-30`, `src/field/boozer_chartmap_io.f90:92-100`, and `python/libneo/chartmap_io.py`. Make it optional on read so existing `.chartmap.nc` files stay valid; guard the writer on `use_B_r`. This touches the format consumed by SIMPLE and rabe, so it must land after step 1-8 and go through the downstream gate on its own.

## Acceptance criteria

- [ ] `make && cd build && ctest --output-on-failure` passes, including the new `test_bcovar_symflux`.
- [ ] `python -m pytest python/tests/test_eqdsk_to_boozer_chartmap.py -vv` passes, including `test_B_theta_matches_plasma_current` (unchanged tolerance `[0.7, 1.05] * 0.2 * |Ip|`).
- [ ] `abs(C_norm - (B_phi_B + B_theta_B/q)) < 1e-10 * abs(C_norm)` at `s in {0.2, 0.5, 0.8}` on `test/resources/input_efit_file.dat`.
- [ ] `Bcov_phi` from splines matches `R*Bp` from a direct `field_eq` call to `1e-10` relative; `Bcov_theta`, `Bcov_s` match `Br*dR_dx + Bz*dZ_dx` to `1e-8` relative.
- [ ] `test/source/test_boozer_converter_vs_simple.f90` is unchanged and still passes (no regression in the VMEC covariant path).
- [ ] `fromefit.bc`, `fromefit_neo_lhs.bc`, `fromefit_neo_rhs.bc` written by `efit_to_boozer.x` are byte-identical to the pre-change output for `test/resources/input_efit_file.dat` (step 1-8 adds an API, it must not move the .bc numbers).
- [ ] The downstream gate (`.github/workflows/downstream-gate.yml`: SIMPLE, NEO-2, MEPHIT, KAMEL, rabe) is green.

## References

- Boozer, A. H., *Plasma equilibrium with rational magnetic surfaces*, Phys. Fluids **24**, 1999 (1981) — definition of the Boozer angles and the `(G + iota*I)` Jacobian.
- Boozer, A. H., *Guiding center drift equations*, Phys. Fluids **23**, 904 (1980) — why the covariant components `I(s)`, `G(s)` and `B_s` are the natural inputs to guiding-centre motion.
- Kruskal, M. D. and Kulsrud, R. M., *Equilibrium of a magnetically confined plasma in a toroid*, Phys. Fluids **1**, 265 (1958) — magnetic coordinates and the flux-function currents.
- In-repo derivation: `doc/efit_to_boozer/EFIT_to_Boozer_and_Hamada_converter.tex`, §"Covariant field components" (lines 230-300), eqs. `(B_r)`, `(B_r_expl)`, `(B_theta)`, `(Brm0num)`.
- Other codes: `booz_xform` (Landreman) writes `bsubumnc`/`bsubvmnc`/`bsubsmns`, i.e. the covariant components, into `boozmn*.nc` alongside `bmnc` — libneo already reads that file in `src/field/boozmn_reader.f90` and could cross-check against it. STELLOPT/BOOZ_XFORM and SFINCS consume exactly these arrays. VMEC's own `wout` carries `bsubsmns`, `bsubumnc`, `bsubvmnc` in VMEC angles, which is the route `src/boozer_converter.F90` already takes via `vmec_field`.

## Scope and risk

Steps 1-8 are one PR of moderate size, additive and behind optional arguments; the main risk is a sign/handedness slip between the `sigma = ±1` left/right-handed branches that `app/efit_to_boozer.f90:100-140` documents, which the `C_norm = G + iota*I` identity check in the acceptance criteria is designed to catch. Step 9 changes an on-disk interchange format consumed by SIMPLE and rabe and must be a separate, gated PR. Related: itpplasma/libneo#360 (clean, COCOS-explicit `efit_to_boozer` rewrite — if that lands first, implement this on top of it rather than on the legacy code), itpplasma/libneo#305 (standalone Boozer evaluator, the natural long-term home for a single shared covariant-component routine), itpplasma/libneo#361 (GPEC comparison, which needs the covariant components to be meaningful).
